### DISCUSSION TITLE
UART blocked communication

### DISCUSSION THREAD

**Discussion by bog739** [micropython/micropython#10403] [Using MicroPython] [answered]
Hi! I am stuck with a problem related to UART on Pi Pico board with an RP2040 MCU. I use the module 0-UART (the default one) to connect to a GSM, everything works just fine when setting up the GSM, for example with text mode. The problem arises when it receives a SMS when I try reading some data the program just freezes. I used uasyncio module to create two tasks, one for constantly checking for data and other which will parse the message if GSM gets one. I used the GSM device on a separate board connected to my laptop and by using classic Python on windows it gets the job done seamlessly. If anyone can help me with an idea, I will be thankful!

**Comment by karfas** [comment]
The exact micropython version you use would be useful.
There were some recent changes to async UART read - see e.g. https://github.com/micropython/micropython/discussions/9951#discussioncomment-4144364.

**Reply by bog739** [reply]
I use micropython 1.19.1 version

**Comment by peterhinch** [comment, marked as answer]
Given that you are using `uasyncio` are you using the `StreamReader` and `StreamWriter` classes? These are highly recommended. See [this example](https://github.com/peterhinch/micropython-async/blob/master/v3/as_demos/auart.py).

**Reply by bog739** [reply]
I used them, but did not use the drain() method after a write. It's a must have?
I'd checked these links before started with uasyncio module:
1. https://github.com/peterhinch/micropython-async/blob/master/v3/docs/TUTORIAL.md#63-using-the-stream-mechanism
2. https://github.com/micropython/micropython/issues/6621
3. https://github.com/peterhinch/micropython-async/blob/master/v3/as_demos/auart_hd.py

**Reply by peterhinch** [reply]
It is necessary to use `.drain()` as per your item 1.

Item 2 is an invalid way to use a stream writer as explained in my response.

Item 3 uses `.awrite()`. This is a legacy mechanism: if you look at [the code](https://github.com/micropython/micropython/blob/699477d12d1aa1cb57ff6cfc3f5dc5e97524dbe6/extmod/uasyncio/stream.py#L177) it uses `.drain()` internally. I guess I should update that example.

I suggest you look at the code to see how the `StreamWriter` works. The `.write()` method simply buffers data. It's `.drain()` that does the actual output. It works this way for CPython compatibility. The mechanism has been subject to discussion and objections, including [from me](https://github.com/micropython/micropython/pull/7868). But there it is.

**Reply by bog739** [reply]
It works now with **`.drain()`** after **`.write()`**. The **`StreamWriter`** and **`StreamReader`** do their job correctly now.
